File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/am_to_amap.py
# Ham chuyen ma tran ke sang adjacency map
def am_to_amap(matrix):
    n = len(matrix)
    amap = {}  # Khoi tao dict trong Python

    for i in range(n):
        for j in range(n):
            count = matrix[i][j]
            if count > 0:
                if i not in amap:
                    amap[i] = {}
                amap[i][j] = count  # Luu so canh tu i den j

    return amap

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
print("Nhap ma tran ke (cach nhau bang dau cach):")
matrix = []
for _ in range(n):
    row = list(map(int, input().split()))
    matrix.append(row)

# Chuyen doi va in ket qua
amap = am_to_amap(matrix)

print("Ban do ke (AMap):")
for u in amap:
    print(f"{u}: ", end="")
    for v, w in amap[u].items():
        print(f"({v},{w})", end=" ")
    print()

File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/am_to_amap.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

// Ham chuyen AM sang AMap
map<int, map<int, int>> amToAmap(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    map<int, map<int, int>> amap; // Su dung map long nhau

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            int count = matrix[i][j];
            if (count > 0) {
                amap[i][j] = count; // Gan so canh vao amap[i][j]
            }
        }
    }

    return amap;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;

    vector<vector<int>> matrix(n, vector<int>(n));
    cout << "Nhap ma tran ke:\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> matrix[i][j];

    // Chuyen doi va in ket qua
    auto amap = amToAmap(matrix);

    cout << "Ban do ke (AMap):\n";
    for (const auto& [u, neighbors] : amap) {
        cout << u << ": ";
        for (const auto& [v, w] : neighbors) {
            cout << "(" << v << "," << w << ") ";
        }
        cout << endl;
    }

    return 0;
}